examples/labs/inverse_kinematics.py

```diff
@@ -0,0 +1,150 @@
+# *********************************************
+
+
+# If jfk is body jacobian, delta_theta is computed by
+#          pose * exp(jfk * delta_theta) = targeted_pose,
+# which has a closed-form solution:
+# .  delta_theta = jfk.pinverse() * log(pose^-1 * targeted_pose)
+# Otherwise, if jfk is spatial jacobian, delta_theta is computed by
+#          exp(jfk * delta_theta) * pose = targeted_pose
+def compute_delta_theta(jfk, theta, targeted_pose, use_body_jacobian):
+    jac, poses = jfk(theta)
+    pose_inv = SE3_Func.inv(poses[-1])
+    error = (
+        SE3_Func.log(
+            SE3_Func.compose(pose_inv, targeted_pose)
+            if use_body_jacobian
+            else SE3_Func.compose(targeted_pose, pose_inv)
+        )
+        .view(-1, 6, 1)
+        .view(-1, 6, 1)
+    )
+    return (jac[-1].pinverse() @ error).view(-1, robot.dof), error.norm().item()
+
+
+# The following function runs IK. We first create random joint angles to compute
+# a set of target poses for the end effector. Then, starting from zero joint angles,
+# we try to recover joint angles that can achieve the target pose, by iteratively
+# updating the joint angles using the body/spatial jacobian. Note that IK has infinite
+# solutions, so, while we are able to recover the target poses after a few iterations,
+# there is no guarantee that we can recover the joint angles used to generate these
+# poses in the first place.
+def run_ik_using_XYZ_method(batch_size, step_size, use_body_jacobian):
+    target_theta = torch.rand(batch_size, robot.dof, dtype=dtype)
+    target_pose: torch.Tensor = fk(target_theta)[0]
+    theta_opt = torch.zeros_like(target_theta)
+    for _ in range(50):
+        delta_theta, error = compute_delta_theta(
+            jfk_b, theta_opt, target_pose, use_body_jacobian
+        )
+        print(error)
+        if error < 1e-4:
+            break
+        theta_opt = theta_opt + step_size * delta_theta
+
+
+print("---------------------------------------------------")
+print("Body Jacobian")
+print("---------------------------------------------------")
+run_ik_using_XYZ_method(100, 0.5, use_body_jacobian=True)
+
+print("---------------------------------------------------")
+print("Spatial Jacobian")
+print("---------------------------------------------------")
+run_ik_using_XYZ_method(10, 0.2, use_body_jacobian=True)
+
+
+# *********************************************
+# ** INVERSE KINEMATICS AS LS OPTIMIZATION
+# *********************************************
+print("---------------------------------------------------")
+print("Theseus Optimizer")
+print("---------------------------------------------------")
+
+
+# IK can also be solved as an optimization problem:
+#      min \|log(pose^-1 * targeted_pose)\|^2
+# as the following
+def targeted_pose_error(optim_vars, aux_vars):
+    (theta,) = optim_vars
+    (targeted_pose,) = aux_vars
+    pose = th.SE3(tensor=fk(theta.tensor)[0])
+    return (pose.inverse().compose(targeted_pose)).log_map()
```

Comment:
we can call `targeted_pose.local(pose)`?

Reply:
Is it `pose.local(targeted_pose)`? In either case, agree that we should call `local()`.

Reply:
That's what we have in the `Local` cost function in /embodied/misc. In the longer term we should actually make a unified version of this cost function that can optionally call in to FK and apply local on pose/point/rot of a specific joint.

Reply:
I think the right way to do this would be [this](https://github.com/facebookresearch/theseus/pull/279#discussion_r955357896), which unfortunately we don't have time to implement, but it would be very nice to have.